src/lib/FreeC/Pass/CompletePatternPass.hs

```diff
@@ -0,0 +1,146 @@
+--
+--   should not pass the check because the @case@ expression is missing an
+--   alternative for the constructor @Nothing@.
+--
+--   == Example 2
+--   The following declaration with redundant alternatives
+--
+--   > redundant :: Just Bool -> Just Bool
+--   > redundant (on :: Just Bool) = case (on :: Just Bool) of {
+--   >     Just a  -> Just False;
+--   >     Nothing -> Nothing;
+--   >     Just b  -> Just True}
+--   >   }
+--
+--   should not pass the check because the @case@ expression has two alternatives
+--   for the constructor @Just@.
+--
+--   == Example 3
+--   The following declaration where the scrutinee is a function
+--
+--   > case_id = case \x -> x  of
+--
+--   should not pass the check because the type of the scrutinee.
+--
+--   = Specification
+--
+--   == Preconditions
+--
+--   The type of all checked expressions has to be annotated.
+--   The Environment has to contain the names of all constructors for
+--   all used data types.
+--   Additionally, the environment should contain entries for all used type
+--   synonyms.
+--
+--   == Translation
+--
+--   This pass only performs a check and therefore does not change the module.
+--   == Postconditions
+--
+--   All case expressions are guaranteed to have complete pattern matching.
+--
+--   == Error cases
+--
+--   A fatal error is reported if an incomplete case expression is found.
+
+module FreeC.Pass.CompletePatternPass
+  ( completePatternPass
+  , checkPatternFuncDecl
+  )
+where
+
+import           Control.Monad                  ( unless )
+import           Data.Maybe                     ( fromJust )
+
+import           FreeC.Environment
+import           FreeC.Environment.Entry
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan
+import           FreeC.IR.TypeSynExpansion
+import           FreeC.Monad.Converter
+import           FreeC.Monad.Reporter
+import           FreeC.Pass
+import           FreeC.Pretty                   ( showPretty )
+
+-- | Checks that all functions of a given module have complete pattern matching.
+--
+--   The pattern matching for a function is complete if for each @case@
+--   expression there exists exactly one case alternative for each constructor
+--   of the corresponding type.
+completePatternPass :: Pass IR.Module
+completePatternPass ast = do
+  mapM_ checkPatternFuncDecl (IR.modFuncDecls ast)
+  return ast
+
+-- | Checks a function declaration for incomplete patterns.
+checkPatternFuncDecl :: IR.FuncDecl -> Converter ()
+checkPatternFuncDecl funcDecl = checkPatternExpr (IR.funcDeclRhs funcDecl)
+ where
+  -- | Checks an expression for incomplete patterns.
+  checkPatternExpr :: IR.Expr -> Converter ()
+  checkPatternExpr (IR.Case srcSpan exprScrutinee exprAlts _) = do
+    checkPatternExpr exprScrutinee
+    mapM_ (checkPatternExpr . IR.altRhs) exprAlts
+    let tau = fromJust $ IR.exprType exprScrutinee -- is safe beacause all types are annotated

```

Comment:
Try to write full sentences in comments. I'd also write the comment before the line. Writing comments at the end of lines just tempts one to write shorter comments (especially due to the 80 character limit).